adapter/src/hermes/adapter/mpiio/mpiio.cc

```diff
@@ -0,0 +1,1006 @@
+      if (read_size != count) {
+        LOG(ERROR) << "reading failed: read " << read_size << " of " << count
+                   << "." << std::endl;
+      }
+    }
+    status = MPI_File_close(&fh);
+  }
+  INTERCEPTOR_LIST->hermes_flush_exclusion.erase(filename);
+  return read_size;
+}
+
+size_t perform_file_write(std::string &filename, int offset, int count,
+                          MPI_Datatype datatype, unsigned char *data_ptr) {
+  LOG(INFO) << "Writing to file: " << filename << " offset: " << offset
+            << " of count:" << count << " datatype:" << datatype << "."
+            << std::endl;
+  INTERCEPTOR_LIST->hermes_flush_exclusion.insert(filename);
+  MPI_File fh;
+  int status = MPI_File_open(MPI_COMM_SELF, filename.c_str(), MPI_MODE_RDWR,
+                             MPI_INFO_NULL, &fh);
+  int write_size = 0;
+  if (fh != nullptr) {
+    status = MPI_File_seek(fh, offset, MPI_SEEK_SET);
+    if (status == 0) {
+      MPI_Status write_status;
+      status = MPI_File_write(fh, data_ptr, count, datatype, &write_status);
+      MPI_Get_count(&write_status, datatype, &write_size);
+      if (write_size != count) {
+        LOG(ERROR) << "writing failed: wrote " << write_size << " of " << count
+                   << "." << std::endl;
+      }
+      status = MPI_File_close(&fh);
+    }
+  }
+  INTERCEPTOR_LIST->hermes_flush_exclusion.erase(filename);
+  return write_size;
+}
+
+std::pair<int, size_t> write_internal(std::pair<AdapterStat, bool> &existing,
+                                      const void *ptr, int count,
+                                      MPI_Datatype datatype, MPI_File *fp,
+                                      MPI_Status *mpi_status,
+                                      bool is_collective = false) {
+  (void)is_collective;
+  LOG(INFO) << "Write called for filename: "
+            << existing.first.st_bkid->GetName()
+            << " on offset: " << existing.first.ptr << " and count: " << count
+            << std::endl;
+  size_t ret;
+  auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+  auto mapper = MapperFactory().Get(kMapperType);
+  int datatype_size;
+  MPI_Type_size(datatype, &datatype_size);
+  size_t total_size = datatype_size * count;
+  auto mapping = mapper->map(FileStruct(fp, existing.first.ptr, total_size));
+  size_t data_offset = 0;
+  auto filename = existing.first.st_bkid->GetName();
+  LOG(INFO) << "Mapping for read has " << mapping.size() << " mapping."
+            << std::endl;
+  for (const auto &item : mapping) {
+    hapi::Context ctx;
+    auto index = std::stol(item.second.blob_name_) - 1;
+    unsigned char *put_data_ptr = (unsigned char *)ptr + data_offset;
+    size_t put_data_ptr_size = item.first.size_;
+
+    if (item.second.size_ == kPageSize) {
+      LOG(INFO) << "Create or Overwrite blob " << item.second.blob_name_
+                << " of size:" << item.second.size_ << "." << std::endl;
+      if (item.second.size_ == kPageSize) {
```

Comment:
Since the `if` conditions on 173 and 176 are the same, doesn't that make lines 185-214 unreachable? In other words, if the condition on 173 is true, then 176 is also true, and we will never hit the `else` branches on 185 and 194.

Reply:
Commented the unreachable code out and created issue #203.